### ISSUE TITLE
Performance regression when creating particles

### ISSUE THREAD

**Issue by RudolfWeeber** [espressomd/espresso#3594]
The following script changed from ~0.8 to 50 secs with recent changes in the Python branch
```
import espressomd
import numpy as np

s=espressomd.System(box_l=[10,10,10])
s.part.add(pos=np.zeros((100000,3)))
```

It is my impression that this is caused in/around the particle node lookup, particularly, when the particle node cache is rebuilt.

In my actual simulations, I need 1^6 particles, so the particle creation would take nearly ten minutes...
It gets worse, when bonds are involved, because then, even more particle node lookups happen.

@fweik, could you take a look, please?
I can probably work around it temporarily as long as my simulations are on a single node, but probably this will come up in more situations.

**Comment by RudolfWeeber** [comment]
With full optimization, perf reports
60% in cell_structure::add_local_particle and 30% in get_maximal_particle_id.

**Comment by fweik** [comment]
From a quick glance the reason is that the python code gets the maximal id after every added particle,
which causes a lot of communication. This is actually not needed, for a slice it only needs to be calculated once, and then all the subsequent ids can be filled in.

**Comment by RudolfWeeber** [comment]
Profile with -Og: 
```
# Event count (approx.): 881886834380
#
# Overhead  Command  Shared Object                                      Symbol                                                                                                                                                                                                                                                                                                                                                                                                                               
# ........  .......  .................................................  .....................................................................................................................................................................................................................................................................................................................................................................................................................................
#
    17.06%  python3  EspressoCore.so                                    [.] CellStructure::update_particle_index
     7.59%  python3  _init.so                                           [.] std::__detail::_Hash_node_value_base<std::pair<int const, int> >::_M_v
     7.50%  python3  EspressoCore.so                                    [.] CellStructure::update_particle_index
     5.75%  python3  EspressoCore.so                                    [.] get_maximal_particle_id()::$_6::operator()
     5.13%  python3  EspressoCore.so                                    [.] std::pair<int, int>::pair<int const, int, true>
     4.98%  python3  _init.so                                           [.] std::max<int>
     4.40%  python3  _init.so                                           [.] __gnu_cxx::__aligned_buffer<std::pair<int const, int> >::_M_addr
     4.22%  python3  ScriptInterface.so                                 [.] std::__detail::_Node_iterator_base<std::pair<int const, int>, false>::_M_incr
     3.82%  python3  _init.so                                           [.] std::__detail::_Hash_node<std::pair<int const, int>, false>::_M_next
     2.77%  python3  ScriptInterface.so                                 [.] std::__detail::operator!=<std::pair<int const, int>, false>
     2.74%  python3  EspressoCore.so                                    [.] std::__detail::operator!=<std::pair<int const, int>, false>@plt
     2.73%  python3  EspressoCore.so                                    [.] std::pair<int, int>::pair<int const, int, true>@plt
     2.26%  python3  EspressoCore.so                                    [.] std::__detail::_Node_const_iterator<std::pair<int const, int>, false, false>::operator*@plt
     2.22%  python3  _init.so                                           [.] __gnu_cxx::__aligned_buffer<std::pair<int const, int> >::_M_ptr
     2.17%  python3  ScriptInterface.so                                 [.] std::__detail::_Node_const_iterator<std::pair<int const, int>, false, false>::operator++
     2.15%  python3  EspressoCore.so                                    [.] std::accumulate<std::__detail::_Node_const_iterator<std::pair<int const, int>, false, false>, int, get_maximal_particle_id()::$_6>
     2.11%  python3  EspressoCore.so                                    [.] std::vector<Particle*, std::allocator<Particle*> >::operator[]
     1.88%  python3  EspressoCore.so                                    [.] std::vector<Particle*, std::allocator<Particle*> >::size
     1.75%  python3  mpiio.so                                           [.] std::__addressof<Particle>
     1.56%  python3  EspressoCore.so                                    [.] Particle::identity
     1.25%  python3  ScriptInterface.so                                 [.] std::__detail::_Node_const_iterator<std::pair<int const, int>, false, false>::operator*
     1.18%  python3  EspressoCore.so                                    [.] std::addressof<Particle>
     0.97%  python3  _init.so                                           [.] std::__detail::_Hash_node_value_base<std::pair<int const, int> >::_M_valptr
     0.84%  python3  EspressoCore.so                                    [.] CellStructure::update_particle_index@plt
     0.55%  python3  EspressoCore.so                                    [.] std::max<int>@plt
     0.52%  python3  ScriptInterface.so                                 [.] std::__detail::_Hash_node<std::pair<int const, int>, false>::_M_next@plt
     0.46%  python3  EspressoCore.so                                    [.] std::vector<Particle*, std::allocator<Particle*> >::operator[]@plt
     0.36%  python3  EspressoCore.so                                    [.] Particle::identity@plt
     0.32%  python3  EspressoCore.so                                    [.] std::vector<Particle*, std::allocator<Particle*> >::size@plt
```

**Comment by fweik** [comment]
In your example script you are adding all particles to the same cell, that also hurts performance.

**Comment by RudolfWeeber** [comment]
I’m working on the suggested change in the Py interface. It halves the problem, but 25sec for 100k particles still isn’t ok.

My actual simulation uses more evenly spaced particles.

**Comment by fweik** [comment]
As I was saying, your example script does add all the particles at the same position. This is not realistic, so this is not a good benchmark.

**Comment by fweik** [comment]
So the background is, every time a cell reallocates, the addresses of all the particle in it changes. So the index has to be updated, which is linear in he number of particles already in the cell. Because the cell reallocates every fixed number of particles added, this makes the whole process quadratic. This has always been the case. The recent changes made the prefactor worse, which can be fixed, but I don't think this situation is worth optimizing for (at least in this way).

**Comment by RudolfWeeber** [comment]
So, for random positions it's somewhat better, but 500k particles still take over a minute.
Things get worse, when adding bonds (or likely other particle properites). Then 100k particles take over a minute.

This is, with the particle id pre-calcualtion, you suggested.

```
import espressomd
import numpy as np


s=espressomd.System(box_l=[10,10,10])
h=espressomd.interactions.HarmonicBond(r_0=0,k=1)
s.bonded_inter.add(h)
n=100000
bonds=[[]]
for i in range(1,n):
  bonds.append([(h,i-1)])
s.part.add(pos=np.random.random((n,3))*s.box_l,bonds=bonds)
```

**Comment by fweik** [comment]
Did you compare to an older version? I understand the index thing, but the performance of the bonds  adding should not have changed recently. I think the proper fix for all of this would be to allow bulk-adding particles...

**Comment by fweik** [comment]
Ah I found something.

**Comment by RudolfWeeber** [comment]
With bonds it takes 1sec with an older version. 
Most of the time in the current code for the bodns case is spentin get_particle_node().

W/r to the index, would it help to reserve capacity in the vector in bigger chunks?

**Comment by fweik** [comment]
I found the problem in `get_particle_node`, just give me a minute. Then I can do a benchmark with more realistic positions.

**Comment by fweik** [comment]
#3595 restores the old behavior as far as I can tell. If you need a faster solution, we should
add a function to bulk-add particles, which would maybe be a good idea in general.

**Comment by fweik** [comment]
```python
s.part.add(pos=np.random.random((100000, 3)) * s.box_l, id=range(100000)) 
```
now takes 1.094s on my machine.

**Comment by fweik** [comment]
Fixed by #3598.
